File: PA59.java
DFS solution
//Leetcode 199. Binary Tree Right Side View
//Time Complexity : O(n)  // n is the number of nodes in the tree
//Space Complexity: O(n) // size of Stack, where size is equal to height of the tree i.e log(n)
class Solution {
    List<Integer> res;
    public List<Integer> rightSideView(TreeNode root) {
        
        res = new ArrayList<>();
        if(root==null){
            return res;
        }
        dfs(root,0);
        return res;
    }
    private void dfs(TreeNode n,int level){
        if(n==null){
            return ;
        }
        if(level==res.size()){
            res.add(n.val);
        }
        dfs(n.right,level+1);
        dfs(n.left,level+1);
        return;
        
        
    }
}

========================================
BFS Solution:
//Leetcode 199. Binary Tree Right Side View
//Time Complexity : O(n)  // n is the number of nodes in the tree
//Space Complexity: O(n) // size of Queue, where size is equal to max no nodes at a level in tree i.e n/2

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        
        List<Integer> res= new ArrayList<>();
        if(root==null){
            return res;
        }
        Queue<TreeNode> q= new LinkedList<>();
        q.offer(root);
        while(!q.isEmpty()){
            int len= q.size();
            
            for(int i=0; i<len;i++){
                TreeNode n= q.poll();
                if(i==len-1){
                    res.add(n.val);
                }
                if(n.left!=null) q.offer(n.left);
                if(n.right!=null) q.offer(n.right);
            }
            
        }
        return res;
        
    }
}
====================================================

File: PA_59.py
#LC 199 - Binary Tree Right Side View
#Time Complexity - O(n)
#Space Complexity - O(h)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        q = deque()
        q.append(root)
        ans = []
        while q:
            l = len(q)
            for i in range(l):
                node = q.popleft()
                if i == l-1:
                    ans.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return ans